x/crosschain/keeper/refund.go

```diff
@@ -0,0 +1,114 @@
+package keeper
+
+import (
+	"errors"
+	"fmt"
+
+	cosmoserrors "cosmossdk.io/errors"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/zeta-chain/zetacore/common"
+	"github.com/zeta-chain/zetacore/x/crosschain/types"
+	fungibletypes "github.com/zeta-chain/zetacore/x/fungible/types"
+	zetaObserverTypes "github.com/zeta-chain/zetacore/x/observer/types"
+)
+
+func (k Keeper) RefundAbortedAmountOnZetaChain(ctx sdk.Context, cctx types.CrossChainTx) error {
+	coinType := cctx.InboundTxParams.CoinType
+	switch coinType {
+	case common.CoinType_Gas:
+		return k.RefundAmountOnZetaChainGas(ctx, cctx)
+	case common.CoinType_Zeta:
+		return k.RefundAmountOnZetaChainZeta(ctx, cctx)
+	case common.CoinType_ERC20:
+		return k.RefundAmountOnZetaChainERC20(ctx, cctx)
+	default:
+		return errors.New("unsupported coin type for refund on ZetaChain")
+	}
+}
+
+// RefundAmountOnZetaChainGas refunds the amount of the cctx on ZetaChain in case of aborted cctx with cointype gas
+func (k Keeper) RefundAmountOnZetaChainGas(ctx sdk.Context, cctx types.CrossChainTx) error {
+	// refund in gas token of a sender chain to the tx origin
+	chainID := cctx.InboundTxParams.SenderChainId
+	amountOfGasTokenLocked := cctx.InboundTxParams.Amount
+	refundTo := ethcommon.HexToAddress(cctx.InboundTxParams.TxOrigin)
+	if refundTo == (ethcommon.Address{}) {
+		return errors.New("invalid address for refund")
+	}
+	if chain := k.zetaObserverKeeper.GetSupportedChainFromChainID(ctx, chainID); chain == nil {
+		return zetaObserverTypes.ErrSupportedChains
+	}
+	fcSenderChain, found := k.fungibleKeeper.GetGasCoinForForeignCoin(ctx, chainID)
+	if !found {
+		return types.ErrForeignCoinNotFound
+	}
+	zrc20 := ethcommon.HexToAddress(fcSenderChain.Zrc20ContractAddress)
+	if zrc20 == (ethcommon.Address{}) {
+		return cosmoserrors.Wrapf(types.ErrForeignCoinNotFound, "zrc20 contract address not found for chain %d", chainID)
+	}
+	// Calcualte amount ot zeta to refund to the tx origin
+	zetaAmount, err := k.fungibleKeeper.QueryUniswapV2RouterGetZetaAmountsIn(ctx, amountOfGasTokenLocked.BigInt(), zrc20)
+	if err != nil {
+		return cosmoserrors.Wrapf(fungibletypes.ErrContractCall, err.Error())
+	}
+	// deposit the amount to the tx orgin instead of receiver as this is a refund
+	if err := k.fungibleKeeper.DepositCoinZeta(ctx, refundTo, zetaAmount); err != nil {
+		return errors.New("failed to refund zeta on ZetaChain" + err.Error())
+	}
+	return nil
+}
+
+// RefundAmountOnZetaChainGas refunds the amount of the cctx on ZetaChain in case of aborted cctx with cointype zeta
+func (k Keeper) RefundAmountOnZetaChainZeta(ctx sdk.Context, cctx types.CrossChainTx) error {
```

Comment:
We would have an issue for Bitcoin chains.
Since the message is sent from admin group, as we discussed, we can consider giving the address to get the refund as an option.